### ISSUE TITLE
Documentation of merge behavior

### ISSUE THREAD

**Issue by dosumis** [ontodev/robot#84]
Where can I find documentation of default merge behaviour in the presence of imports statements?  In tests it seems to automatically merge down the import chain.  This can be useful behaviour in some situations, but I don't think it should be the default.

**Comment by cmungall** [comment]
ideally here:

```
$ robot merge -h
usage: robot filter --input <file> --input <file> --output <file>
 -h,--help              print usage information
 -i,--input <arg>       merge ontology from a file
 -I,--input-iri <arg>   merge ontology from an IRI
 -noprefixes            do not use default prefixes
 -o,--output <arg>      save merged ontology to a file
 -p,--prefix <arg>      add a prefix 'foo: http://bar'
 -P,--prefixes <arg>    use prefixes from JSON-LD file
 -v,--version           print version information
```

but this lacks the docs. It also says "filter"...

To avoid docs getting out of sync, how about:

 1. minimal docs on command line, most auto-derived
 2. formal specification in javadocs for the <FOO>Operation.java object
 3. A link from 1 to HTML generated from 2

Alternatively, a markdown or md/wiki page for each operation?

**Comment by jamesaoverton** [comment]
Yes, the merge command merges all imports. Anything else will get complicated fast.

The "filter" mistake is easy to fix.

There's also the examples/README.md documentation, which is automatically tested.

**Comment by dosumis** [comment]
> Yes, the merge command merges all imports. Anything else will get complicated fast.

I think we need to deal with the complexity.

I was trying to merge in some auto-generated files to GO.  It is essential to preserve GO import structure when doing this. OWLtools allows this, so I've gone back to using it.  Merging down import chains is a separate command. 

Keeping these steps separate is also essential to the release pipelines on VFB and FlyBase ontologies.

**Comment by cmungall** [comment]
Attempting to specify current behavior. It's important to distinguish between Operation and Command. There are actually two methods in MergeOperation it's important to distinguish between.

## merge ontologies into target (mergeInto method)

https://github.com/ontodev/robot/blob/6049ab427cc2455cb9c4d14281a254310be9e5b3/robot-core/src/main/java/org/obolibrary/robot/MergeOperation.java#L106-L133

This does not touch the import closure of targetOntology. It does not merge in its imports closure, nor does it remove the declarations themselves.

It does copy the axioms from the full imports closures of the list of other ontologies into target.

This seems reasonable to me. I don't really see a use case for not copying the axioms from the imports closure of the ontologies that are merged in. There is potentially a use case for copying the import directives over to the target (as a distinct option). This corresponds to owltools "merge-support-ontologies"

## Merge command (merge method)

https://github.com/ontodev/robot/blob/6049ab427cc2455cb9c4d14281a254310be9e5b3/robot-core/src/main/java/org/obolibrary/robot/MergeOperation.java#L118-L133

This is a convenience wrapper onto `mergeInto(T, O1, ..., On)`, and corresponds to the command line command.

This threw me at first:

        OWLOntology ontology = ontologies.get(0);
        mergeInto(ontologies, ontology, false);

So we nominate the first in the list as the target, and we also pass the target as one of the ontologies to merge in. So we end up merging the ontology into itself.

This is semantics-preserving so on one level it's OK. But this is surely unintentional @jamesaoverton? Perhaps you were thinking closure and car/cdr...?

It has the IMO undesirable consequence of copying in imported axioms, and preserving the imports declaration. So the same axiom will be the ontology twice. This can be seen in a test ontology 'import_test.owl' that imports simple.owl from the test resources dir.

<img width="1401" alt="screen shot 2016-07-16 at 4 04 03 pm" src="https://cloud.githubusercontent.com/assets/50745/16897741/152da3c2-4b6f-11e6-80be-d567b116da9e.png">

I'd like to refactor this and add options so that we get a range of behaviors for our different use cases, but I want to make sure this won't break your usage James

## Proposed new behavior

I propose 3 distinct behaviors. All behaviors have the following properties:

 * semantics preserving: the results should be an ontology with the same semantics as the combined set of input ontologies
 * does not introduce redundancies (the current behavior fails on this)

In each case, a list is passed in on the command line. The first is designate target, and the second onwards are designated 'remaining' ontologies. The target ontology is mutated, and this is what is saved afterwards.

The behaviors are:

 1. partial merge: preserve imports structure of target (first) ontology, copy all axioms that directly belong remaining ontologies, __plus axioms from the import closure of remaining ontologies__. No analog in owltools.
 2. partial merge: preserve imports structure of target (first) ontology, copy all axioms that direcly belong to remaining ontologies, __copy import directives from remaining into target ontology__. Corresponds to `merge-support-ontologies` in owltools.
 3. full merge:  Copy all axioms that directly belong to remaining ontologies into target; copy axioms from __all__ ontology imports closures into target. Remove imports directives from target. Corresponds to the composition of `merge-support-ontologies o merge-imports-closure` in owltools.

I propose making 3 the default. This seems to be closest to what the current implementation is aiming at.

**Comment by jamesaoverton** [comment]
I think this is fixed by #92.
